R-package/tests/testthat/test_basic.R

```diff
@@ -3652,6 +3652,28 @@ test_that("lgb.cv() only prints eval metrics when expected to", {
   )
 })
 
+test_that("lgb.cv() works with an already constructed dataset with integer categoricals", {
+  data("mtcars")
+  y <- mtcars$mpg
+  x <- as.matrix(mtcars[, -1L])
+  categorical_feature <- which(names(mtcars) %in% c("cyl", "vs", "am", "gear", "carb")) - 1L
+  dtrain <- lgb.Dataset(
+    data = x
+    , label = y
+    , categorical_feature = categorical_feature
+    , free_raw_data = TRUE
+    , params = list(num_threads = .LGB_MAX_THREADS)
+  )
+  dtrain$construct()
+  params <- list(
+    objective = "regression"
+    , num_leaves = 2L
+    , verbose = .LGB_VERBOSITY
+    , num_threads = .LGB_MAX_THREADS
+  )
+  lgb.cv(params = params, data = dtrain, nrounds = 1L)
```

Comment:
could you add an assertion here that something happened? Just to give us a bit more confidence that this behavior didn't result in something like an early return with no training.

For example, maybe some of these checks that the returned objective contains metrics computed on the validation sets created by `lgb.cv()`:

https://github.com/microsoft/LightGBM/blob/6e78e697f8e27d4bbf493a0ebd9134c3254a2ecd/R-package/tests/testthat/test_basic.R#L469-L492